airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java

```diff
@@ -0,0 +1,341 @@
+              .collect(Collectors.toList()));
+    }
+  }
+
+  protected List<String> getExcludedInternalSchemas() {
+    return List.of();
+  }
+
+  protected List<TableInfo> getTables(final Database database) throws Exception {
+    return discoverInternal(database).stream()
+        .map(t -> {
+          final List<Field> fields = Arrays.stream(t.fields())
+              .map(f -> Field.of(f.getName(), jooqDataTypeToJsonSchemaType(f.getDataType())))
+              .collect(Collectors.toList());
+
+          return new TableInfo(String.format("%s.%s", t.getSchema().getName(), t.getName()), fields);
+        })
+        .collect(Collectors.toList());
+  }
+
+  private List<Table<?>> discoverInternal(final Database database) throws Exception {
+    return database.query(context -> {
+      final List<Schema> schemas = context.meta().getSchemas();
+      return schemas.stream()
+          .filter(schema -> !getExcludedInternalSchemas().contains(schema.getName()))
+          .flatMap(schema -> context.meta(schema).getTables().stream())
+          .collect(Collectors.toList());
+    });
+  }
+
+  @Override
+  public Stream<AirbyteMessage> read(JsonNode config, ConfiguredAirbyteCatalog catalog, JsonNode state) throws Exception {
+    final Instant now = Instant.now();
+
+    final Database database = createDatabase(config);
+
+    final Map<String, Table<?>> tableNameToTable = discoverInternal(database).stream()
+        .collect(Collectors.toMap(t -> String.format("%s.%s", t.getSchema().getName(), t.getName()), Function.identity()));
+
+    Stream<AirbyteMessage> resultStream = Stream.empty();
+
+    for (final ConfiguredAirbyteStream airbyteStream : catalog.getStreams()) {
+      final String streamName = airbyteStream.getStream().getName();
+      if (!tableNameToTable.containsKey(streamName)) {
+        continue;
+      }
+
+      final Set<String> selectedFields = CatalogHelpers.getTopLevelFieldNames(airbyteStream);
+
+      final Table<?> table = tableNameToTable.get(streamName);
+      final List<org.jooq.Field<?>> selectedDatabaseFields = Arrays.stream(table.fields())
+          .filter(field -> selectedFields.contains(field.getName()))
+          .collect(Collectors.toList());
+
+      if (selectedDatabaseFields.isEmpty()) {
+        continue;
+      }
+
+      final Stream<AirbyteMessage> stream;
+      if (airbyteStream.getSyncMode() == SyncMode.INCREMENTAL) {
+        final JdbcStateManager stateManager = new JdbcStateManager(state == null ? new JdbcState() : Jsons.object(state, JdbcState.class));
+        final String cursorField = IncrementalUtils.getCursorField(airbyteStream);
+        final JsonSchemaPrimitive cursorType = IncrementalUtils.getCursorType(airbyteStream, cursorField);
+        final Optional<String> initialCursorOptional = stateManager.getOriginalCursor(streamName);
+
+        final Stream<AirbyteMessage> internalMessageStream;
+        if (initialCursorOptional.isPresent()) {
```

Comment:
the logic right now seems to be: if there was ever a cursor, do incremental. else do fullrefresh. 

Shouldn't it be the other way around i.e: if there is a current cursor AND that cursor is the same as the initial cursor do incremental otherwise FR? might be missing something.

Reply:
i think you're right. i'll double check this.